components/script/dom/selection.rs

```diff
@@ -0,0 +1,522 @@
+            Err(Error::InvalidState)
+        }
+    }
+
+    // https://w3c.github.io/selection-api/#dom-selection-extend
+    // TODO: When implementing actual selection UI, this may be the correct
+    // method to call as the continue-selection action
+    fn Extend(&self, node: &Node, offset: u32) -> ErrorResult {
+        if !self.is_same_root(node) {
+            // Step 1
+            return Ok(());
+        }
+        if let Some(range) = self.range.get() {
+            if node.is_doctype() {
+                return Err(Error::InvalidNodeType);
+            }
+
+            if offset > node.len() {
+                // As with is_doctype, not explicit in selection spec steps
+                // implied by range-setting
+                return Err(Error::IndexSize);
+            }
+
+            // let is_old_anchor_before_or_equal;
+            // Step 4
+            if !self.is_same_root(&*range.StartContainer()) {
+                // Step 5, and its following 8 and 9
+                self.set_range(&*Range::new(&self.document, node, offset, node, offset));
+                self.direction.set(Direction::Forwards);
+            } else {
+                let old_anchor_node = &*self.GetAnchorNode().unwrap(); // has range, therefore has anchor node
+                let old_anchor_offset = self.AnchorOffset();
+                let is_old_anchor_before_or_equal = {
+                    if old_anchor_node == node {
+                        old_anchor_offset <= offset
+                    } else {
+                        old_anchor_node.is_before(node)
+                    }
+                };
+                if is_old_anchor_before_or_equal {
+                    // Step 6, and its following 8 and 9
+                    self.set_range(&*Range::new(
+                        &self.document,
+                        old_anchor_node,
+                        old_anchor_offset,
+                        node,
+                        offset,
+                    ));
+                    self.direction.set(Direction::Forwards);
+                } else {
+                    // Step 7, and its following 8 and 9
+                    self.set_range(&*Range::new(
+                        &self.document,
+                        node,
+                        offset,
+                        old_anchor_node,
+                        old_anchor_offset,
+                    ));
+                    self.direction.set(Direction::Backwards);
+                }
+            };
+        } else {
+            // Step 2
+            return Err(Error::InvalidState);
+        }
+        return Ok(());
+    }
+
+    // https://w3c.github.io/selection-api/#dom-selection-setbaseandextent
+    fn SetBaseAndExtent(
+        &self,
+        anchor_node: &Node,
+        anchor_offset: u32,
+        focus_node: &Node,
+        focus_offset: u32,
+    ) -> ErrorResult {
+        // Step 1
+        if anchor_node.is_doctype() || focus_node.is_doctype() {
```

Comment:
Presumably this check comes from https://dom.spec.whatwg.org/#concept-range-bp-set? I'm a bit concerned that we're adding these extra checks in various places in the code; is our implementation missing them elsewhere, or is this a spec inconsistency, or are we working around tests that assume a certain order or checks that doesn't match what the spec says?

Reply:
Tests, which Firefox and Chrome both pass in FYI, are assuming a certain order that doesn't match what the spec says.

Reply:
I brought it up here: https://github.com/w3c/selection-api/issues/118